src/test/java/com/testapp/subject/controller/SubjectControllerIntegrationTest.java

```diff
@@ -0,0 +1,174 @@
+        mockMvc
+                .perform(MockMvcRequestBuilders.get("/subjects"))
+                .andExpect(MockMvcResultMatchers.status().isOk());
+    }
+
+    @Test
+    public void when_user_is_not_authenticated_in_add_page_it_should_redirect_to_login() throws Exception{
+        mockMvc
+                .perform(MockMvcRequestBuilders.get("/subject/add"))
+                .andExpect(MockMvcResultMatchers.redirectedUrl("http://localhost/login"));
+    }
+
+    @Test
+    @WithMockUser(username = "test", password = "test")
+    public void when_user_is_authenticated_in_add_page_it_should_be_display() throws Exception{
+        mockMvc
+                .perform(MockMvcRequestBuilders.get("/subject/add"))
+                .andExpect(MockMvcResultMatchers.status().isOk());
+    }
+
+    @Test
+    @WithMockUser(username = "test", password = "test")
+    public void when_subject_created_it_should_show_in_the_list() throws Exception{
+        mockMvc
+                .perform(MockMvcRequestBuilders.post("/subject/add")
+                        .param("title","math72"))
+                .andExpect(MockMvcResultMatchers.redirectedUrl("/subjects"))
+                .andDo(
+                        result -> MockMvcResultMatchers
+                                .xpath("//table//tbody//tr[0]//td[1]//a[contains(text(), 'math72')]")
+                                .exists());
+      Optional<Subject> subjects = subjectRepository.findFirstByTitle("math72");
+        subjectRepository.delete(subjects.get());
+
+    }
+    @Test
+    @WithMockUser(username = "test", password = "test")
+    public void when_title_is_empty_create_form_should_display_validation_error() throws Exception{
+        mockMvc
+                .perform(MockMvcRequestBuilders.post("/subject/add")
+                        .param("title", " "))
+                .andExpect(
+                        result ->  MockMvcResultMatchers
+                        .xpath("//div[contains(text(), 'subject  cannot be empty')]")
+                        .exists());
+    }
+
+    @Test
+    @WithMockUser(username = "test", password = "test")
+    public void when_title_already_exists_create_form_should_display_validation_error() throws Exception{
+          Subject subject = new Subject();
+          subject.setTitle("Dummy");
+          subjectRepository.save(subject);
+          mockMvc
+                  .perform(MockMvcRequestBuilders.post("/subject/add")
+                          .param("title", "Dummy"))
+                  .andExpect(
+                          result -> MockMvcResultMatchers
+                                  .xpath("//div[contains(text(), 'subject should be unique')]")
+                                  .exists());
+          subjectRepository.delete(subject);
+    }
+
+    @Test
+    public void when_user_is_not_authenticated_edit_page_it_should_redirect_to_login() throws  Exception{
+        Subject subject = new Subject();
+        subject.setTitle("math94");
```

Comment:
why do you need to create subject if you want to test redirection to login page? 🤔 

please also remove the subjectRepository.delete(subject); from line 112